<h1 align=center><font size = 5>Build a Regression Model in Keras</font></h1> 

In this course project, you will build a regression model using the deep learning Keras library, and then you will experiment with increasing the number of training epochs and changing number of hidden layers and you will see how changing these parameters impacts the performance of the model.

##  Getting our Dataset Ready

We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

Let's download the data and read it into a <em>pandas</em> dataframe.

In [2]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

#### Let's check how many data points we have.

In [3]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

#### Split data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Let's save the number of predictors to *n_cols* since we will need this number when building our network.

In [10]:
n_cols = predictors_norm.shape[1] # number of predictors

## Import Keras

Let's go ahead and import the Keras library

In [11]:
import keras

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

Let's import the rest of the packages from the Keras library that we will need to build our regression model.

In [12]:
from keras.models import Sequential
from keras.layers import Dense

## Build a Neural Network

Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

In [13]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model that has one hidden layers consisting of 10 hidden units. The network uses <strong>reLU</strong> as the activation function and <strong>mean squared error</strong> as the loss function

## Train and Test the Network

Let's call the function now to create our model.

In [14]:
model = regression_model()

We will import the <strong>train_test_split</strong> function from <strong>scikit-learn</strong> to randomly split the data into a training and test sets. Also, we will import the <strong>mean_squared_error</strong> function from Scikit-learn.

In [15]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

Next, We will create a function called <strong>evaluate_model</strong> which will perform the following tasks.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing

2. Train and test the model at the same time using the <strong>fit</strong> method. We will train the model for 50 epochs.

3. After training is complete, We will need to evaluate the model on the test data.

4. Compute the mean squared error between the predicted concrete strength and the actual concrete strength.

5. Return the mean_squared_error

In [16]:
def evaluate_model(predictors, target, test_size = 0.3, random_state = 1, epochs = 50, verbose = 1):
    # Split data 
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=test_size, random_state=random_state)
    # Fit the model
    model.fit(X_train, y_train, epochs=epochs, verbose=verbose)
    # Evaluate model
    loss_val = model.evaluate(X_test, y_test,verbose=verbose)
    y_pred = model.predict(X_test)
    #Compute the mean squared error
    mean_square_error = mean_squared_error(y_test, y_pred)
    return (loss_val, mean_square_error)

In [17]:
print(evaluate_model(predictors, target,0.3,47,100,1))

Epoch 1/100
721/721 [==============================] - 3s 4ms/step - loss: 1534.9242
Epoch 2/100
721/721 [==============================] - 2s 2ms/step - loss: 1518.7255
Epoch 3/100
721/721 [==============================] - 1s 1ms/step - loss: 1502.8610
Epoch 4/100
721/721 [==============================] - 1s 1ms/step - loss: 1487.1528
Epoch 5/100
721/721 [==============================] - 1s 2ms/step - loss: 1471.4846
Epoch 6/100
721/721 [==============================] - 1s 2ms/step - loss: 1455.7013: 0s - loss: 146
Epoch 7/100
721/721 [==============================] - 1s 2ms/step - loss: 1440.0077
Epoch 8/100
721/721 [==============================] - 1s 1ms/step - loss: 1423.6131
Epoch 9/100
721/721 [==============================] - 1s 2ms/step - loss: 1407.0023
Epoch 10/100
721/721 [==============================] - 1s 2ms/step - loss: 1390.0370
Epoch 11/100
721/721 [==============================] - 1s 2ms/step - loss: 1372.4515
Epoch 12/100
721/721 [=========================

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/keras/callbacks.py:120: UserWarning: Method on_batch_end() is slow compared to the batch update (0.139778). Check your callbacks.
  % delta_t_median)


721/721 [==============================] - 1s 1ms/step - loss: 160.4905
Epoch 87/100
721/721 [==============================] - 1s 2ms/step - loss: 159.3293
Epoch 88/100
721/721 [==============================] - 1s 1ms/step - loss: 158.0401
Epoch 89/100
721/721 [==============================] - 1s 2ms/step - loss: 156.8974
Epoch 90/100
721/721 [==============================] - 1s 2ms/step - loss: 155.7295
Epoch 91/100
721/721 [==============================] - 1s 2ms/step - loss: 154.6973
Epoch 92/100
721/721 [==============================] - 1s 2ms/step - loss: 153.5300
Epoch 93/100
721/721 [==============================] - 1s 2ms/step - loss: 152.4562
Epoch 94/100
721/721 [==============================] - 1s 2ms/step - loss: 151.4115
Epoch 95/100
721/721 [==============================] - 1s 1ms/step - loss: 150.3774
Epoch 96/100
721/721 [==============================] - 1s 1ms/step - loss: 149.4378A: 0s - loss: 152.6
Epoch 97/100
721/721 [==============================] - 1s 

Finally, Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [18]:
total_mean_squared_errors = 50
epochs = 100
mean_squared_errors = []

for i in range(0, total_mean_squared_errors):
    MSE,mean_square_error = evaluate_model(predictors, target,0.3,i,epochs,0)
    print("MSE "+str(i+1)+": "+str(MSE))
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Number of epochs for each training: "+str(epochs) )
print("Mean of "+str(total_mean_squared_errors)+" MSE: " + str(mean))
print("Standard Deviation of "+str(total_mean_squared_errors)+" MSE: " + str(standard_deviation))

MSE 1: 90.60787687332498
MSE 2: 74.10165654648469
MSE 3: 43.463698958116055
MSE 4: 39.99970625596525
MSE 5: 40.89263277763688
MSE 6: 41.30647079072724
MSE 7: 43.091546734559884
MSE 8: 31.994773938817886
MSE 9: 36.44973819078365
MSE 10: 31.521122447881112
MSE 11: 36.11074972924291
MSE 12: 32.072723252873594
MSE 13: 39.198150906362194
MSE 14: 41.42523883461566
MSE 15: 33.37205908445093
MSE 16: 29.52399411247772
MSE 17: 33.053259834116716
MSE 18: 34.040116590975174
MSE 19: 31.47845204671224
MSE 20: 35.67805598237368
MSE 21: 32.129110996777186
MSE 22: 31.882980957771967
MSE 23: 28.595760863961527
MSE 24: 33.45597713433423
MSE 25: 33.833442194176335
MSE 26: 36.81785543991138
MSE 27: 30.952346172147585
MSE 28: 31.42083818009756
MSE 29: 36.67698345986771
MSE 30: 37.1819106722341
MSE 31: 32.19212049959547
MSE 32: 30.490356402103956
MSE 33: 31.588490452195447
MSE 34: 32.914674937917965
MSE 35: 35.02432897870209
MSE 36: 39.17050797422341
MSE 37: 27.3425360960482
MSE 38: 36.054574318302485
MSE 39